ouroboros-network/src/Ouroboros/Network/PeerSharing.hs

```diff
@@ -104,11 +115,99 @@ peerSharingClient controlMessageSTM
         putMVar resultQueue result
         peerSharingClient controlMessageSTM psc
 
-peerSharingServer :: Monad m
-                  => (PeerSharingAmount -> m [peer])
+peerSharingServer :: ( MonadSTM m
+                     , MonadMonotonicTime m
+                     , Hashable peer
+                     , RandomGen s
+                     )
+                  => PeerSharingAPI peer s m
                   -> PeerSharingServer peer m
-peerSharingServer computePeersToShare =
+peerSharingServer psAPI =
   PeerSharingServer
-    { recvMsgShareRequest = \amount -> (,) <$> computePeersToShare amount
-                                          <*> return (peerSharingServer computePeersToShare)
+    { recvMsgShareRequest = \amount -> (,) <$> computePeerSharingPeers psAPI amount
+                                          <*> return (peerSharingServer psAPI)
     }
+
+--
+-- Utility Function
+--
+
+-- | PeerSharingAPI needed to compute the peers to be shared.
+--
+data PeerSharingAPI addr s m =
+  PeerSharingAPI { psPublicPeerSelectionStateVar :: StrictTVar m (PublicPeerSelectionState addr)
+                 , psGenVar                      :: StrictTVar m s
```

Comment:
Is there any benefit for `s`? We'll anyway just use `StdGen` (which can be used in both `IO` and `IOSim`), isn't it?

Reply:
unfortunately `s` is needed due to some polymorphism in the test suite.